read tombstones: EOF after upgrade to 0.14.0~n201605160800

We jsut upgraded from 0.12.1-1 to the latest nightly (0.14.0~n201605160800)

On restart, the performance was great, but it seems that only data in the last ~20 hours is acceissible. The other data is still on disk.

There are a bunch of startup errors:

```
[root@carf-metrics-influx01 influxdb]# grep err influxd.log
[store] 2016/05/16 16:14:55 Failed to open shard: 1: [shard 1] error opening memory map for file /data/influxdb-data/metrics/tg_udp/default/1/000002485-000000032.tsm: init: read tombstones: EOF
[store] 2016/05/16 16:14:55 Failed to open shard: 61: [shard 61] error opening memory map for file /data/influxdb-data/metrics/tg_udp/default/61/000000001-000000001.tsm: init: read tombstones: EOF
[store] 2016/05/16 16:14:56 Failed to open shard: 20: [shard 20] error opening memory map for file /data/influxdb-data/metrics/tg_udp/default/20/000003221-000000041.tsm: init: read tombstones: EOF
[store] 2016/05/16 16:14:57 Failed to open shard: 14: [shard 14] error opening memory map for file /data/influxdb-data/metrics/ryan_test/default/14/000000001-000000001.tsm: init: read tombstones: EOF
[store] 2016/05/16 16:14:58 Failed to open shard: 9: [shard 9] error opening memory map for file /data/influxdb-data/metrics/tg_udp/default/9/000003226-000000030.tsm: init: read tombstones: EOF
[store] 2016/05/16 16:14:59 Failed to open shard: 64: [shard 64] error opening memory map for file /data/influxdb-data/metrics/tg_udp/default/64/000000001-000000001.tsm: init: read tombstones: EOF
[store] 2016/05/16 16:14:59 Failed to open shard: 68: [shard 68] error opening memory map for file /data/influxdb-data/metrics/tg_udp/default/68/000000160-000000007.tsm: init: read tombstones: EOF
[store] 2016/05/16 16:18:09 Failed to open shard: 14: [shard 14] error opening memory map for file /data/influxdb-data/metrics/ryan_test/default/14/000000001-000000001.tsm: init: read tombstones: EOF
[store] 2016/05/16 16:18:10 Failed to open shard: 1: [shard 1] error opening memory map for file /data/influxdb-data/metrics/tg_udp/default/1/000002485-000000032.tsm: init: read tombstones: EOF
[store] 2016/05/16 16:18:10 Failed to open shard: 68: [shard 68] error opening memory map for file /data/influxdb-data/metrics/tg_udp/default/68/000000640-000000007.tsm: init: read tombstones: EOF
[store] 2016/05/16 16:18:10 Failed to open shard: 9: [shard 9] error opening memory map for file /data/influxdb-data/metrics/tg_udp/default/9/000003226-000000029.tsm: init: read tombstones: EOF
[store] 2016/05/16 16:18:11 Failed to open shard: 64: [shard 64] error opening memory map for file /data/influxdb-data/metrics/tg_udp/default/64/000000001-000000001.tsm: init: read tombstones: EOF
[store] 2016/05/16 16:18:13 Failed to open shard: 20: [shard 20] error opening memory map for file /data/influxdb-data/metrics/tg_udp/default/20/000003221-000000041.tsm: init: read tombstones: EOF
[store] 2016/05/16 16:18:13 Failed to open shard: 61: [shard 61] error opening memory map for file /data/influxdb-data/metrics/tg_udp/default/61/000000001-000000001.tsm: init: read tombstones: EOF
```

These files seem legit, have the correct permissions, and have a mtime from months ago:

```
[root@carf-metrics-influx01 influxdb]# ls -lh /data/influxdb-data/metrics/tg_udp/default/1/000002485-000000032.tsm
-rw-r--r-- 1 influxdb influxdb 940M Feb  2 07:18 /data/influxdb-data/metrics/tg_udp/default/1/000002485-000000032.tsm
[root@carf-metrics-influx01 influxdb]# ls -lh /data/influxdb-data/metrics/tg_udp/default/61/000000001-000000001.tsm
-rw-r--r-- 1 influxdb influxdb 6.3M Mar 11 19:00 /data/influxdb-data/metrics/tg_udp/default/61/000000001-000000001.tsm
[root@carf-metrics-influx01 influxdb]# ls -lh /data/influxdb-data/metrics/tg_udp/default/20/000003221-000000041.tsm
-rw-r--r-- 1 influxdb influxdb 1.5G Feb 16 09:04 /data/influxdb-data/metrics/tg_udp/default/20/000003221-000000041.tsm
```

Suggestions?

Right, that gets rid of the errors. Sadly now we panic immediately on startup (before any query executes):

```
[retention] 2016/05/16 16:40:58 Starting retention policy enforcement service with check interval of 10m0s
[udp] 2016/05/16 16:40:58 Started listening on UDP: :8089
[udp] 2016/05/16 16:40:58 Started listening on UDP: :8090
[run] 2016/05/16 16:40:58 Listening for signals
panic: interface conversion: tsm1.Value is *tsm1.FloatValue, not *tsm1.IntegerValue

goroutine 1852 [running]:
panic(0xb76bc0, 0xc936def740)
	/usr/local/go/src/runtime/panic.go:481 +0x3e6
github.com/influxdata/influxdb/tsdb/engine/tsm1.encodeIntegerBlock(0x0, 0x0, 0x0, 0xc95ec7c000, 0x2ff, 0x400, 0x0, 0x0, 0x0, 0x0, ...)
	/root/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/encoding.go:771 +0x72b
github.com/influxdata/influxdb/tsdb/engine/tsm1.Values.Encode(0xc95ec7c000, 0x2ff, 0x400, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/root/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/encoding.go:195 +0x22c
github.com/influxdata/influxdb/tsdb/engine/tsm1.(*tsmKeyIterator).chunk(0xc8ca12ef00, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/root/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/compact.go:992 +0x4af
github.com/influxdata/influxdb/tsdb/engine/tsm1.(*tsmKeyIterator).combine(0xc8ca12ef00, 0xc986cd9400, 0x0, 0x0, 0x0)
	/root/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/compact.go:965 +0xd0a
github.com/influxdata/influxdb/tsdb/engine/tsm1.(*tsmKeyIterator).merge(0xc8ca12ef00)
	/root/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/compact.go:845 +0x13f
github.com/influxdata/influxdb/tsdb/engine/tsm1.(*tsmKeyIterator).Next(0xc8ca12ef00, 0xc83e5d0531)
	/root/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/compact.go:813 +0xbe5
github.com/influxdata/influxdb/tsdb/engine/tsm1.(*Compactor).write(0xc822aa8ab0, 0xc926539630, 0x4a, 0x7fe690183820, 0xc8ca12ef00, 0x0, 0x0)
	/root/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/compact.go:560 +0x53b
github.com/influxdata/influxdb/tsdb/engine/tsm1.(*Compactor).writeNewFiles(0xc822aa8ab0, 0x212, 0x5, 0x7fe690183820, 0xc8ca12ef00, 0x0, 0x0, 0x0, 0x0, 0x0)
	/root/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/compact.go:507 +0x3a1
github.com/influxdata/influxdb/tsdb/engine/tsm1.(*Compactor).compact(0xc822aa8ab0, 0xc8222e2500, 0xc9a352dfc0, 0x2, 0x2, 0x0, 0x0, 0x0, 0x0, 0x0)
	/root/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/compact.go:473 +0x4fe
github.com/influxdata/influxdb/tsdb/engine/tsm1.(*Compactor).CompactFull(0xc822aa8ab0, 0xc9a352dfc0, 0x2, 0x2, 0x0, 0x0, 0x0, 0x0, 0x0)
	/root/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/compact.go:478 +0x71
github.com/influxdata/influxdb/tsdb/engine/tsm1.(*Engine).compactTSMFull.func1(0xc966b63f60, 0xc820124580, 0x0, 0xc9a352dfc0, 0x2, 0x2)
	/root/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/engine.go:760 +0x4c6
created by github.com/influxdata/influxdb/tsdb/engine/tsm1.(*Engine).compactTSMFull
	/root/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/engine.go:778 +0x27e
```

New issue? Any suggestions?

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -32,6 +32,7 @@
 - [#6406](https://github.com/influxdata/influxdb/issues/6406): Max index entries exceeded
 - [#6557](https://github.com/influxdata/influxdb/issues/6557): Overwriting points on large series can cause memory spikes during compactions
 - [#6611](https://github.com/influxdata/influxdb/issues/6611): Queries slow down hundreds times after overwriting points
+- [#6641](https://github.com/influxdata/influxdb/issues/6641): Fix read tombstones: EOF
 
 ## v0.13.0 [2016-05-12]
 

--- a/tsdb/engine/tsm1/tombstone.go
+++ b/tsdb/engine/tsm1/tombstone.go
@@ -161,7 +161,10 @@ func (t *Tombstoner) readTombstone() ([]Tombstone, error) {
 		var b [4]byte
 		_, err := f.Read(b[:])
 		if err != nil {
-			return nil, err
+			// Might be a zero length file which should not exist, but
+			// an old bug allowed them to occur.  Treat it as an empty
+			// v1 tombstone file so we don't abort loading the TSM file.
+			return t.readTombstoneV1(f)
 		}
 
 		if _, err := f.Seek(0, os.SEEK_SET); err != nil {
@@ -171,7 +174,6 @@ func (t *Tombstoner) readTombstone() ([]Tombstone, error) {
 		if binary.BigEndian.Uint32(b[:]) == v2header {
 			return t.readTombstoneV2(f)
 		}
-
 		return t.readTombstoneV1(f)
 	}
 	return nil, nil

--- a/tsdb/engine/tsm1/tombstone_test.go
+++ b/tsdb/engine/tsm1/tombstone_test.go
@@ -173,3 +173,31 @@ func TestTombstoner_ReadV1(t *testing.T) {
 		t.Fatalf("value mismatch: got %v, exp %v", got, exp)
 	}
 }
+
+func TestTombstoner_ReadEmptyV1(t *testing.T) {
+	dir := MustTempDir()
+	defer func() { os.RemoveAll(dir) }()
+
+	f := MustTempFile(dir)
+	f.Close()
+
+	if err := os.Rename(f.Name(), f.Name()+".tombstone"); err != nil {
+		t.Fatalf("rename tombstone failed: %v", err)
+	}
+
+	ts := &tsm1.Tombstoner{Path: f.Name()}
+
+	entries, err := ts.ReadAll()
+	if err != nil {
+		fatal(t, "ReadAll", err)
+	}
+
+	entries, err = ts.ReadAll()
+	if err != nil {
+		fatal(t, "ReadAll", err)
+	}
+
+	if got, exp := len(entries), 0; got != exp {
+		t.Fatalf("length mismatch: got %v, exp %v", got, exp)
+	}
+}

